### PR TITLE
Mutation testing

### PR DESCRIPTION
**Pull request by JekaMas**
# Description

As far as consensus is extremely sensitive functionality, I believe we need a way to check our tests for usefulness and correctness. Code coverage, unfortunately, doesn't tell us about the quality of tests, either about usefulness. Even with high code coverage it's possible to write not very useful tests that always 'green', or tests that are 'green' if being executed in particular order.
For that reason the PR introduces changes:
1. As far as we run tests with t.Parallel() it makes sense to use `-race` flag as well. At the moment tests do have data races that'd be great to fix in separate issues.
2. To prevent incorrect global state the PR introduces `-shuffle` flag
3. To check that tests are useful the PR adds mutation testing https://github.com/avito-tech/go-mutesting

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [x] I have added sufficient documentation in code

# Comments
This PR is for discussion about new tooling. At the moment data race tests and mutation tests are `red`. If we'd like to merge the PR, we should comment out this CI steps and make tasks to fix tests with race and a bunch of tasks to add/check tests with respect of mutation testing results.

A few examples of mutation testing:

```
 func (i *IBFT) handlePrepare(view *proto.View, quorum uint64) bool {
 	isValidPrepare := func(message *proto.Message) bool {
 		// Verify that the proposal hash is valid
@@ -799,7 +799,7 @@
 
 	if len(prepareMessages) < int(quorum)-1 {
 		//	quorum not reached, keep polling
-		return false
+
 	}
 
func (i *IBFT) handlePrepare(view *proto.View, quorum uint64) bool {
 	isValidPrepare := func(message *proto.Message) bool {
 		// Verify that the proposal hash is valid
@@ -852,7 +852,7 @@
 		case <-sub.SubCh:
 			if !i.handleCommit(view, quorum) {
 				//	quorum not reached, retry
-				continue
+
 			}
 

 func (i *IBFT) handleCommit(view *proto.View, quorum uint64) bool {
 	isValidCommit := func(message *proto.Message) bool {
 		var (
@@ -880,7 +880,7 @@
 	commitMessage
[result.txt](https://github.com/0xPolygon/go-ibft/files/9723038/result.txt)
s := i.messages.GetValidMessages(view, proto.MessageType_COMMIT, isValidCommit)
 	if len(commitMessages) < int(quorum) {
 		//	quorum not reached, keep polling
-		return false
+
 	}
 


 func (i *IBFT) runStates(ctx context.Context) {
 	var timeout error
 
@@ -594,7 +594,7 @@
 
 	//	is proposer
 	if !i.backend.IsProposer(msg.From, height, round) {
-		return false
+
 	}

 func (i *IBFT) runStates(ctx context.Context) {
 	var timeout error
 
@@ -615,7 +615,7 @@
 
 	//	proposal must be for round 0
 	if msg.View.Round != 0 {
-		return false
+
 	}
 
 func (i *IBFT) handlePrePrepare(view *proto.View) *proto.Message {
 	isValidPrePrepare := func(message *proto.Message) bool {
 		if view.Round == 0 {
@@ -772,7 +772,7 @@
 		case <-sub.SubCh:
 			if !i.handlePrepare(view, quorum) {
 				//	quorum of valid prepare messages not received, retry
-				continue
+
 			}
 
```

More results are in the file [result.txt](https://github.com/0xPolygon/go-ibft/files/9723039/result.txt)

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -62,9 +62,44 @@ jobs:
...
         with:
           files: coverage.out
+
+  mutating:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Install Go
+        uses: actions/setup-go@v3
+        with:
+          go-version: 1.17.x
```
Comment (vcastellm) [inline .github/workflows/main.yml:78]: nitpick: Should be 1.18 already

**Review by vcastellm — COMMENTED** [review]
1. ✅ 
2. ✅ 
3. Let's discuss

**Comment by zivkovicmilos** [comment]
@JekaMas 
I'll review this PR along with @dbrajovic when it's ready, but just one note since you're still working on it:
Please don't drop the `consensus_test.go` file, and move those tests to `ibft_test.go` 🙏 

There is a specific reason why we kept the 2 separate, as `ibft_test.go` contains _only_ unit tests, whereas `consensus_test.go` contains _cluster_ tests

File: .github/workflows/main.yml (hunk 1 above)
Comment (vcastellm) [reply .github/workflows/main.yml:78]: well https://github.com/0xPolygon/go-ibft/blob/main/go.mod#L3 off by a few hours haha

**Comment by JekaMas** [comment]
> @JekaMas I'll review this PR along with @dbrajovic when it's ready, but just one note since you're still working on it: Please don't drop the `consensus_test.go` file, and move those tests to `ibft_test.go` 🙏
> 
> There is a specific reason why we kept the 2 separate, as `ibft_test.go` contains _only_ unit tests, whereas `consensus_test.go` contains _cluster_ tests

I with I could keep this files separated. I left a comment, why this is needed. Otherwise, mutations won't be performed versus consensus_test

**Comment by zivkovicmilos** [comment]
Hey @JekaMas,

Resolved the races in tests in the following PR which I'll merge out to `main`:
PR #29 

I've included your test race workflow there, so when you pull in the changes after it gets merged, you'll have them running

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -62,9 +62,44 @@ jobs:
...
+        with:
+          submodules: recursive
+
+      - name: Install dependencies
+        run: make install
+
+      - name: Mutating testing
+        run: make mut
+
+#  test-race:
```
Comment (vcastellm) [inline .github/workflows/main.yml:91]: Why not use the `-race` flag in the existing workflow? https://github.com/0xPolygon/go-ibft/pull/28/files#diff-7829468e86c1cc5d5133195b5cb48e1ff6c75e3e9203777f6b2e379d9e4882b3R65

File: Makefile [hunk 3]
```diff
@@ -1,6 +1,13 @@
-.PHONY: lint
+.PHONY: lint mut
 lint:
 	golangci-lint run -E whitespace -E wsl -E wastedassign -E unconvert -E tparallel -E thelper -E stylecheck -E prealloc \
 	-E predeclared -E nlreturn -E misspell -E makezero -E lll -E importas -E ifshort -E gosec -E  gofmt -E goconst \
 	-E forcetypeassert -E dogsled -E dupl -E errname -E errorlint -E nolintlint --timeout 2m
 
+install:
```
Comment (zivkovicmilos) [inline Makefile:7]: What do you think about renaming this to `install-mut`, so it's more clear?

File: Makefile [hunk 4]
```diff
@@ -1,6 +1,13 @@
...
+	go get -u github.com/avito-tech/go-mutesting/...
+	go install github.com/avito-tech/go-mutesting/...
+
+mut:
+	go-mutesting --blacklist=".github/mut_blacklist" --config=".github/mut_config.yml" ./...
+	@echo MSI: `jq '.stats.msi' report.json`
```
Comment (vcastellm) [inline Makefile:13]: Nitpick: Missing new line

**Comment by zivkovicmilos** [comment]
@JekaMas 
The comment just states that the files need to be joined up, and I can't find anything on the official docs about it.
Can you please explain the reasoning because I'm probably missing something?

**Comment by zivkovicmilos** [comment]
@JekaMas @vcastellm 
Are we sure we want to add this mutating workflow, since it runs for basically 36m?

It greatly increases the runtime for PR checks.

cc @epikichi @ZeljkoBenovic

**Comment by vcastellm** [comment]
> @JekaMas @vcastellm Are we sure we want to add this mutating workflow, since it runs for [basically 36m](https://github.com/0xPolygon/go-ibft/actions/runs/3227586302/jobs/5282585559)?
> 
> It greatly increases the runtime for PR checks.
> 
> cc @epikichi @ZeljkoBenovic

Good question, can we reduce the scope in the config somehow? Does it provide enough value that justifies this 36m?

**Comment by JekaMas** [comment]
> > @JekaMas @vcastellm Are we sure we want to add this mutating workflow, since it runs for [basically 36m](https://github.com/0xPolygon/go-ibft/actions/runs/3227586302/jobs/5282585559)?
> > It greatly increases the runtime for PR checks.
> > cc @epikichi @ZeljkoBenovic
> 
> Good question, can we reduce the scope in the config somehow? Does it provide enough value that justifies this 36m?

Could we have separate tasks to make mutation testing like
1. full run for releases
2. only for a PR changes (tests) for the PR?

**Comment by zivkovicmilos** [comment]
> Sure, it is from the mutation testing package readme https://github.com/avito-tech/go-mutesting
The targets of the mutation testing can be defined as arguments to the binary. Every target can be either a Go source file, a directory or a package. Directories and packages can also include the ... wildcard pattern which will search recursively for Go source files. Test source files with the suffix _test are excluded, since this would interfere with the testing process most of the time.
Also, I just tried it.

Isn't it possible to just give the directory / package for mutation, instead of a single test file? I've ran it with `./...`

**Comment by JekaMas** [comment]
> @JekaMas
> 
> > Sure, it is from the mutation testing package readme https://github.com/avito-tech/go-mutesting
> > The targets of the mutation testing can be defined as arguments to the binary. Every target can be either a Go source file, a directory or a package. Directories and packages can also include the ... wildcard pattern which will search recursively for Go source files. Test source files with the suffix _test are excluded, since this would interfere with the testing process most of the time.
> > Also, I just tried it.
> 
> Isn't it possible to just give the directory / package for mutation, instead of a single test file? I've ran it with `./...`

It is possible, although it'd run file by file:
1. greb a file `file.go`
2. perform a mutation
3. run tests `file_test.go`
4. mark the result

**Comment by JekaMas** [comment]
@zivkovicmilos Could you check last changes? I tried different approach with test files.

**Comment by zivkovicmilos** [comment]
@JekaMas 
>@zivkovicmilos Could you check the last changes? I tried a different approach with test files.

I like this idea of having a test that will be the cornerstone for the cluster mutation tests 💯 
It's much cleaner now

**Comment by JekaMas** [comment]
closed in favor of https://github.com/0xPolygon/go-ibft/pull/38

**Pull request closed without merging**